src/value_generators/include/value_generators/value_generators.hpp

```diff
@@ -69,6 +77,262 @@ class DocumentGenerator {
+ */
+class Expression {
+public:
+    using Parser = std::function<UniqueExpression(YAML::Node)>;
+
+    /**
+     * Parse `node` into one of the Expression types registered in the `parserMap`.
+     *
+     * @throws InvalidConfigurationException if `node` isn't a mapping type with a '^'-prefixed key.
+     */
+    static UniqueExpression parseExpression(YAML::Node node);
+
+    /**
+     * Parse `node` into either:
+     *   - one of the Expression types registered in the `parserMap`, or
+     *   - a DocumentExpression instance containing further nested Expression types.
+     *
+     * @throws InvalidConfigurationException if `node` isn't a mapping type.
+     */
+    static UniqueExpression parseObject(YAML::Node node);
+
+    /**
+     * Parse `node` into one of the following:
+     *   - one of the Expression types registered in the `parserMap`,
+     *   - a DocumentExpression instance containing further nested Expression types,
+     *   - an ArrayExpression instance containing a list Expression types,
+     *   - a ConstantExpression instance representing a scalar value.
+     *
+     * @throws InvalidConfigurationException if `node` is malformed.
+     */
+    static UniqueExpression parseOperand(YAML::Node node);
+
+    virtual ~Expression() = default;
+
+    /**
+     * Generate a possibly randomized value.
+     *
+     * @throws InvalidConfigurationException if the arguments to this Expression evaluate an
+     * unexpected type.
+     */
+    virtual Value evaluate(genny::DefaultRandom& rng) const = 0;
+};
+
+
+/**
+ * Class for representing a fixed scalar value.
+ */
+class ConstantExpression : public Expression {
+public:
+    static UniqueExpression parse(YAML::Node node);
+
+    ConstantExpression(Value value);
+    Value evaluate(genny::DefaultRandom& rng) const override;
+
+private:
+    const Value _value;
+};
+
+
+/**
+ * Class for generating a bsoncxx::document::value with its elements generated from other Expression
+ * types.
+ */
+class DocumentExpression : public Expression {
+public:
+    using ElementType = std::pair<std::string, UniqueExpression>;
+
+    static UniqueExpression parse(YAML::Node node);
+
+    DocumentExpression(std::vector<ElementType> elements);
+    Value evaluate(genny::DefaultRandom& rng) const override;
+
+private:
+    const std::vector<ElementType> _elements;
+};
+
+
+/**
+ * Class for generating a bsoncxx::array::value with its elements generated from other Expression
+ * types.
+ */
+class ArrayExpression : public Expression {
+public:
+    using ElementType = UniqueExpression;
+
+    static UniqueExpression parse(YAML::Node node);
+
+    ArrayExpression(std::vector<ElementType> elements);
+    Value evaluate(genny::DefaultRandom& rng) const override;
+
+private:
+    const std::vector<ElementType> _elements;
+};
+
+
+/**
+ * Base class for generating random int64_t values.
+ */
+class RandomIntExpression : public Expression {
+public:
+    static UniqueExpression parse(YAML::Node node);
```

Comment:
> (Or, rather, can these be moved into `genny::v1` and a simple "use this function to extract value-generators" provided?)

Maybe. I had considered making `Expression::parseExpression()` and `Expression::parseObject()` `protected` functions so they could only be used when defining `Expression` subclasses, but then it seemed somewhat arbitrary those functions wouldn't be part of the public API and yet `ArrayExpression::parse()` would.

We could make all but `Expression::parseOperand()` `protected` functions if you don't feel strongly about having callers use a different one to require a certain type. (We could also add some other scheme to the `Expression` class in the future to learn the type without `evaluate()`ing a `Value` for it.) I didn't attempt to hide much of the implementation because all `Expression` subclasses need to be added to `parserMap` until we add a more dynamic registration scheme akin to `Cast::Registration`.